File: app/src/main/java/com/apps/adrcotfas/goodtime/AddEditLabels/AddEditLabelActivity.java
/*
 * Copyright 2016-2019 Adrian Cotfas
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License. You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License is
 * distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
 * either express or implied. See the License for the specific language governing permissions and limitations under the License.
 */

package com.apps.adrcotfas.goodtime.AddEditLabels;

import android.content.Context;
import android.os.Bundle;
import android.view.MenuItem;
import android.view.View;
import android.view.inputmethod.EditorInfo;
import android.widget.EditText;
import android.widget.FrameLayout;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.TextView;
import android.widget.Toast;

import com.apps.adrcotfas.goodtime.BL.GoodtimeApplication;
import com.apps.adrcotfas.goodtime.BL.PreferenceHelper;
import com.apps.adrcotfas.goodtime.Label;
import com.apps.adrcotfas.goodtime.Main.LabelsViewModel;
import com.apps.adrcotfas.goodtime.Main.SimpleItemTouchHelperCallback;
import com.apps.adrcotfas.goodtime.R;
import com.apps.adrcotfas.goodtime.Util.ThemeHelper;
import com.apps.adrcotfas.goodtime.databinding.ActivityAddEditLabelsBinding;
import com.google.android.material.snackbar.BaseTransientBottomBar;
import com.google.android.material.snackbar.Snackbar;
import com.takisoft.colorpicker.ColorPickerDialog;

import java.util.List;

import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import androidx.core.content.ContextCompat;
import androidx.databinding.DataBindingUtil;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.ViewModelProviders;
import androidx.recyclerview.widget.DefaultItemAnimator;
import androidx.recyclerview.widget.ItemTouchHelper;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import static com.apps.adrcotfas.goodtime.Util.ThemeHelper.COLOR_INDEX_UNLABELED;
import static com.apps.adrcotfas.goodtime.Util.ThemeHelper.clearFocusEditText;
import static com.apps.adrcotfas.goodtime.Util.ThemeHelper.requestFocusEditText;

public class AddEditLabelActivity extends AppCompatActivity
        implements AddEditLabelsAdapter.OnEditLabelListener{

    private static final String TAG = AddEditLabelActivity.class.getSimpleName();

    private LabelsViewModel mLabelsViewModel;
    private List<Label> mLabels;

    private RecyclerView mRecyclerView;
    private AddEditLabelsAdapter mCustomAdapter;
    private ItemTouchHelper mItemTouchHelper;

    private Label mLabelToAdd;

    private LinearLayout mEmptyState;
    private EditText mAddLabelView;
    private FrameLayout mImageRightContainer;
    private ImageView mImageLeft;
    private FrameLayout mImageLeftContainer;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        ThemeHelper.setTheme(this);

        ActivityAddEditLabelsBinding binding = DataBindingUtil.setContentView(this, R.layout.activity_add_edit_labels);
        mLabelsViewModel = ViewModelProviders.of(this).get(LabelsViewModel.class);

        setSupportActionBar(binding.toolbarWrapper.toolbar);
        if (getSupportActionBar() != null) {
            getSupportActionBar().setDisplayHomeAsUpEnabled(true);
        }

        mRecyclerView = binding.labelList;
        mEmptyState = binding.emptyState;
        mAddLabelView = binding.addLabel.text;
        mImageRightContainer = binding.addLabel.imageRightContainer;
        mImageLeft = binding.addLabel.imageLeft;
        mImageLeftContainer = binding.addLabel.imageLeftContainer;

        final LiveData<List<Label>> labelsLiveData = mLabelsViewModel.getAllLabels();
        labelsLiveData.observe(this, labels -> {

            mLabels = labels;

            mCustomAdapter = new AddEditLabelsAdapter(this, mLabels, this);
            mRecyclerView.setAdapter(mCustomAdapter);

            LinearLayoutManager lm = new LinearLayoutManager(this);
            lm.setReverseLayout(true);
            lm.setStackFromEnd(true);
            mRecyclerView.setLayoutManager(lm);

            mRecyclerView.setItemAnimator(new DefaultItemAnimator());

            labelsLiveData.removeObservers(AddEditLabelActivity.this);

            binding.progressBar.setVisibility(View.GONE);
            updateRecyclerViewVisibility();

            ItemTouchHelper.Callback callback = new SimpleItemTouchHelperCallback(mCustomAdapter);
            mItemTouchHelper = new ItemTouchHelper(callback);
            mItemTouchHelper.attachToRecyclerView(mRecyclerView);
        });

        mLabelToAdd = new Label("", COLOR_INDEX_UNLABELED);

        mImageRightContainer.setOnClickListener(view -> {
            addLabel();
            updateRecyclerViewVisibility();
            clearFocusEditText(mAddLabelView, this);
        });

        mImageLeftContainer.setOnClickListener(v -> requestFocusEditText(mAddLabelView, this));

        mAddLabelView.setOnFocusChangeListener((view, hasFocus) -> {
            mImageRightContainer.setVisibility(hasFocus ? View.VISIBLE : View.INVISIBLE);
            mImageLeft.setImageDrawable(getResources().getDrawable(hasFocus ? R.drawable.ic_palette : R.drawable.ic_add));
            mImageLeftContainer.setOnClickListener(hasFocus ? v -> {
                final ColorPickerDialog.Params p = new ColorPickerDialog.Params.Builder(AddEditLabelActivity.this)
                        .setColors(ThemeHelper.getPalette(this))
                        .setSelectedColor(ThemeHelper.getColor(this, mLabelToAdd.colorId))
                        .build();
                ColorPickerDialog dialog = new ColorPickerDialog(AddEditLabelActivity.this, R.style.DialogTheme, c
                        -> {
                    mLabelToAdd.colorId = ThemeHelper.getIndexOfColor(this, c);
                    mImageLeft.setColorFilter(c);
                }, p);
                dialog.setTitle(R.string.label_select_color);
                dialog.show();
            } : v -> requestFocusEditText(mAddLabelView, this));
            mImageLeft.setColorFilter(ThemeHelper.getColor(this, hasFocus ? mLabelToAdd.colorId : COLOR_INDEX_UNLABELED) );
            if (!hasFocus) {
                mLabelToAdd = new Label("", ThemeHelper.getColor(this, COLOR_INDEX_UNLABELED));
                mAddLabelView.setText("");
            }
        });

        mAddLabelView.setOnEditorActionListener((v, actionId, event) -> {
            if(actionId == EditorInfo.IME_ACTION_DONE){
                addLabel();
                updateRecyclerViewVisibility();
                clearFocusEditText(mAddLabelView, this);
                return true;
            }
            return false;
        });
    }

    @Override
    public void onDragStarted(RecyclerView.ViewHolder viewHolder) {
        mItemTouchHelper.startDrag(viewHolder);
    }

    @Override
    public void onEditLabel(String label, String newLabel) {
        mLabelsViewModel.editLabelName(label, newLabel);

        Label crtLabel = PreferenceHelper.getCurrentSessionLabel();
        if (crtLabel.title != null && crtLabel.title.equals(label)) {
            PreferenceHelper.setCurrentSessionLabel(new Label(newLabel, crtLabel.colorId));
        }
    }

    @Override
    public void onEditColor(String label, int color) {
        mLabelsViewModel.editLabelColor(label, color);

        Label crtLabel = PreferenceHelper.getCurrentSessionLabel();
        if (crtLabel.title != null && crtLabel.title.equals(label)) {
            PreferenceHelper.setCurrentSessionLabel(new Label(label, color));
        }
    }

    @Override
    public void onToggleArchive(Label label, int adapterPosition) {
        mLabelsViewModel.toggleLabelArchive(label.title, label.archived);

        Label crtLabel = PreferenceHelper.getCurrentSessionLabel();
        if (label.archived && crtLabel.title != null && crtLabel.title.equals(label.title)) {
            GoodtimeApplication.getCurrentSessionManager().getCurrentSession().setLabel(null);
            PreferenceHelper.setCurrentSessionLabel(new Label(null, COLOR_INDEX_UNLABELED));
        }
        if (label.archived && !PreferenceHelper.getArchivedLabelHintWasShown()) {
            showArchivedLabelHint();
        }
    }

    /**
     * When archiving a label for the first time, a snackbar will be shown to explain the action.
     */
    private void showArchivedLabelHint() {
        Snackbar s = Snackbar.make(mRecyclerView, getString(R.string.tutorial_archive_label), Snackbar.LENGTH_INDEFINITE)
                .setAction("OK", view -> {
                    PreferenceHelper.setArchivedLabelHintWasShown(true);
                })
                .setActionTextColor(getResources().getColor(R.color.teal200));

        s.setBehavior(new BaseTransientBottomBar.Behavior() {
            @Override
            public boolean canSwipeDismissView(View child) {
                return false;
            }
        });
        TextView tv = s.getView().findViewById(com.google.android.material.R.id.snackbar_text);
        if (tv != null) {
            tv.setTextColor(ContextCompat.getColor(this, R.color.white));
        }
        s.show();
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case android.R.id.home:
                finish();
                return true;
        }
        return super.onOptionsItemSelected(item);
    }

    @Override
    public void onDeleteLabel(Label label, int position) {
        mLabels.remove(label);
        mCustomAdapter.notifyItemRemoved(position);
        mLabelsViewModel.deleteLabel(label.title);

        // workaround for edge case: clipping animation of last cached entry
        if (mCustomAdapter.getItemCount() == 0) {
            mCustomAdapter.notifyDataSetChanged();
        }

        String crtSessionLabel = GoodtimeApplication.getCurrentSessionManager().getCurrentSession().getLabel().getValue();
        if (crtSessionLabel != null && crtSessionLabel.equals(label.title)) {
            GoodtimeApplication.getCurrentSessionManager().getCurrentSession().setLabel(null);
        }

        // the label attached to the current session was deleted
        if (label.title.equals(PreferenceHelper.getCurrentSessionLabel().title)) {
            PreferenceHelper.setCurrentSessionLabel(new Label(null, COLOR_INDEX_UNLABELED));
        }

        updateRecyclerViewVisibility();
    }

    /**
     * Update the order of the labels inside the database based on the
     * rearrangement that was done inside the adapter.
     */
    @Override
    public void onLabelRearranged() {
        for (int i = 0; i < mLabels.size(); ++i) {
            mLabelsViewModel.editLabelOrder(mLabels.get(i).title, i);
        }
    }

    private void updateRecyclerViewVisibility() {
        if (mCustomAdapter.getItemCount() == 0) {
            mRecyclerView.setVisibility(View.GONE);
            mEmptyState.setVisibility(View.VISIBLE);
        }
        else {
            mRecyclerView.setVisibility(View.VISIBLE);
            mEmptyState.setVisibility(View.GONE);
        }
    }

    /**
     * Used for checking label names before adding or renaming existing ones.
     * Checks for invalid strings like empty ones, spaces or duplicates
     * @param newLabel The desired name for the new label or renamed label
     * @return true if the label name is valid, false otherwise
     */
    public static boolean labelIsGoodToAdd(Context context, List<Label> labels, String newLabel, String beforeEdit) {
        boolean result = true;

        if (beforeEdit != null && beforeEdit.equals(newLabel)) {
            result = false;
        } else if (newLabel.length() == 0) {
            result = false;
        } else {
            boolean duplicateFound = false;
            for (Label l : labels) {
                if (newLabel.equals(l.title)) {
                    duplicateFound = true;
                    break;
                }
            }
            if (duplicateFound) {
                Toast.makeText(context, R.string.label_already_exists, Toast.LENGTH_SHORT).show();
                result = false;
            }
        }
        return result;
    }

    private void addLabel() {
        mLabelToAdd = new Label(mAddLabelView.getText().toString().trim(), mLabelToAdd.colorId);
        if (labelIsGoodToAdd(this, mLabels, mLabelToAdd.title, null)) {
            mLabels.add(mLabelToAdd);

            mCustomAdapter.notifyItemInserted(mLabels.size());
            mRecyclerView.scrollToPosition(mLabels.size() - 1);

            mLabelsViewModel.addLabel(mLabelToAdd);
            mLabelToAdd = new Label("", COLOR_INDEX_UNLABELED);
            mAddLabelView.setText("");
        }
    }
}

File: app/src/main/java/com/apps/adrcotfas/goodtime/labels/AddEditLabelActivity.kt
/*
 * Copyright 2016-2021 Adrian Cotfas
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License. You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License is
 * distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
 * either express or implied. See the License for the specific language governing permissions and limitations under the License.
 */
package com.apps.adrcotfas.goodtime.labels

import android.content.Context
import android.os.Bundle
import android.view.KeyEvent
import android.view.MenuItem
import android.view.View
import android.view.View.OnFocusChangeListener
import android.view.inputmethod.EditorInfo
import android.widget.*
import androidx.activity.viewModels
import androidx.appcompat.app.AppCompatActivity
import androidx.databinding.DataBindingUtil
import androidx.recyclerview.widget.DefaultItemAnimator
import androidx.recyclerview.widget.ItemTouchHelper
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.apps.adrcotfas.goodtime.R
import com.apps.adrcotfas.goodtime.bl.CurrentSessionManager
import com.apps.adrcotfas.goodtime.database.Label
import com.apps.adrcotfas.goodtime.databinding.ActivityAddEditLabelsBinding
import com.apps.adrcotfas.goodtime.labels.AddEditLabelsAdapter.OnEditLabelListener
import com.apps.adrcotfas.goodtime.main.LabelsViewModel
import com.apps.adrcotfas.goodtime.main.SimpleItemTouchHelperCallback
import com.apps.adrcotfas.goodtime.settings.PreferenceHelper
import com.apps.adrcotfas.goodtime.statistics.Utils.getInvalidLabelWithRandomColor
import com.apps.adrcotfas.goodtime.util.ThemeHelper
import com.google.android.material.snackbar.BaseTransientBottomBar
import com.google.android.material.snackbar.Snackbar
import com.takisoft.colorpicker.ColorPickerDialog
import dagger.hilt.android.AndroidEntryPoint
import javax.inject.Inject

@AndroidEntryPoint
class AddEditLabelActivity : AppCompatActivity(), OnEditLabelListener {

    private val viewModel: LabelsViewModel by viewModels()
    private lateinit var labels: MutableList<Label>
    private lateinit var recyclerView: RecyclerView
    private lateinit var adapter: AddEditLabelsAdapter
    private lateinit var itemTouchHelper: ItemTouchHelper
    private lateinit var labelToAdd: Label
    private lateinit var emptyState: LinearLayout
    private lateinit var addLabelView: EditText
    private lateinit var imageRight: FrameLayout
    private lateinit var imageLeft: ImageView
    private lateinit var imageLeftContainer: FrameLayout

    @Inject
    lateinit var preferenceHelper: PreferenceHelper

    @Inject
    lateinit var currentSessionManager: CurrentSessionManager

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        ThemeHelper.setTheme(this, preferenceHelper.isAmoledTheme())
        val binding: ActivityAddEditLabelsBinding =
            DataBindingUtil.setContentView(this, R.layout.activity_add_edit_labels)
        setSupportActionBar(binding.toolbarWrapper.toolbar)
        supportActionBar?.setDisplayHomeAsUpEnabled(true)
        recyclerView = binding.labelList
        emptyState = binding.emptyState
        addLabelView = binding.addLabel.text
        imageRight = binding.addLabel.imageRightContainer
        imageLeft = binding.addLabel.imageLeft
        imageLeftContainer = binding.addLabel.imageLeftContainer
        val labelsLiveData = viewModel.allLabels
        labelsLiveData.observe(this) { labels: List<Label> ->
            this.labels = labels as MutableList<Label>
            adapter = AddEditLabelsAdapter(this, this.labels, this)
            recyclerView.adapter = adapter
            val lm = LinearLayoutManager(this)
            lm.reverseLayout = true
            lm.stackFromEnd = true
            recyclerView.layoutManager = lm
            recyclerView.itemAnimator = DefaultItemAnimator()
            labelsLiveData.removeObservers(this@AddEditLabelActivity)
            binding.progressBar.visibility = View.GONE
            updateRecyclerViewVisibility()
            val callback: ItemTouchHelper.Callback = SimpleItemTouchHelperCallback(adapter)
            itemTouchHelper = ItemTouchHelper(callback)
            itemTouchHelper.attachToRecyclerView(recyclerView)
        }
        labelToAdd = getInvalidLabelWithRandomColor(this)
        imageRight.setOnClickListener {
            addLabel()
            updateRecyclerViewVisibility()
            ThemeHelper.clearFocusEditText(addLabelView, this)
        }
        imageLeftContainer.setOnClickListener {
            ThemeHelper.requestFocusEditText(
                addLabelView,
                this
            )
        }
        addLabelView.onFocusChangeListener =
            OnFocusChangeListener { _: View?, hasFocus: Boolean ->
                imageRight.visibility = if (hasFocus) View.VISIBLE else View.INVISIBLE
                imageLeft.setImageDrawable(resources.getDrawable(if (hasFocus) R.drawable.ic_palette else R.drawable.ic_add))
                imageLeftContainer.setOnClickListener(if (hasFocus) View.OnClickListener {
                    val p = ColorPickerDialog.Params.Builder(this@AddEditLabelActivity)
                        .setColors(ThemeHelper.getPalette(this))
                        .setSelectedColor(ThemeHelper.getColor(this, labelToAdd.colorId))
                        .build()
                    val dialog = ColorPickerDialog(
                        this@AddEditLabelActivity,
                        R.style.DialogTheme,
                        { c: Int ->
                            labelToAdd.colorId = ThemeHelper.getIndexOfColor(this, c)
                            imageLeft.setColorFilter(c)
                        },
                        p
                    )
                    dialog.setTitle(R.string.label_select_color)
                    dialog.show()
                } else View.OnClickListener {
                    ThemeHelper.requestFocusEditText(
                        addLabelView,
                        this
                    )
                })
                imageLeft.setColorFilter(
                    ThemeHelper.getColor(
                        this,
                        if (hasFocus) labelToAdd.colorId else ThemeHelper.COLOR_INDEX_UNLABELED
                    )
                )
                if (!hasFocus) {
                    labelToAdd = getInvalidLabelWithRandomColor(this)
                    addLabelView.setText("")
                }
            }
        addLabelView.setOnEditorActionListener { _: TextView?, actionId: Int, _: KeyEvent? ->
            if (actionId == EditorInfo.IME_ACTION_DONE) {
                addLabel()
                updateRecyclerViewVisibility()
                ThemeHelper.clearFocusEditText(addLabelView, this)
                return@setOnEditorActionListener true
            }
            false
        }
    }

    override fun onDragStarted(viewHolder: RecyclerView.ViewHolder) {
        itemTouchHelper.startDrag(viewHolder)
    }

    override fun onEditLabel(label: String, newLabel: String) {
        viewModel.editLabelName(label, newLabel)
        val crtLabel = preferenceHelper.currentSessionLabel
        if (crtLabel.title == label) {
            preferenceHelper.currentSessionLabel = Label(newLabel, crtLabel.colorId)
        }
    }

    override fun onEditColor(label: String, newColor: Int) {
        viewModel.editLabelColor(label, newColor)
        val crtLabel = preferenceHelper.currentSessionLabel
        if (crtLabel.title != "" && crtLabel.title == label) {
            preferenceHelper.currentSessionLabel = Label(label, newColor)
        }
    }

    override fun onToggleArchive(label: Label, adapterPosition: Int) {
        viewModel.toggleLabelArchive(label.title, label.archived)
        val crtLabel = preferenceHelper.currentSessionLabel
        if (label.archived && crtLabel.title != "" && crtLabel.title == label.title) {
            currentSessionManager.currentSession.setLabel("")
            preferenceHelper.currentSessionLabel = Label(
                "", ThemeHelper.getColor(
                    this,
                    ThemeHelper.COLOR_INDEX_UNLABELED
                )
            )
        }
        if (label.archived && !preferenceHelper.archivedLabelHintWasShown) {
            showArchivedLabelHint()
        }
    }

    /**
     * When archiving a label for the first time, a snackbar will be shown to explain the action.
     */
    private fun showArchivedLabelHint() {
        val s = Snackbar.make(
            recyclerView,
            getString(R.string.tutorial_archive_label),
            Snackbar.LENGTH_INDEFINITE
        )
            .setAction(getString(android.R.string.ok)) { preferenceHelper.archivedLabelHintWasShown = true }
        s.behavior = object : BaseTransientBottomBar.Behavior() {
            override fun canSwipeDismissView(child: View): Boolean {
                return false
            }
        }
        s.show()
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        if (item.itemId == android.R.id.home) {
            finish()
            return true
        }
        return super.onOptionsItemSelected(item)
    }

    override fun onDeleteLabel(label: Label, position: Int) {
        labels.remove(label)
        adapter.notifyItemRemoved(position)
        viewModel.deleteLabel(label.title)

        // workaround for edge case: clipping animation of last cached entry
        if (adapter.itemCount == 0) {
            adapter.notifyDataSetChanged()
        }
        val crtSessionLabel = currentSessionManager.currentSession.label.value
        if (crtSessionLabel != null && crtSessionLabel == label.title) {
            currentSessionManager.currentSession.setLabel("")
        }

        // the label attached to the current session was deleted
        if (label.title == preferenceHelper.currentSessionLabel.title) {
            preferenceHelper.currentSessionLabel = Label(
                "", ThemeHelper.getColor(
                    this,
                    ThemeHelper.COLOR_INDEX_UNLABELED
                )
            )
        }
        updateRecyclerViewVisibility()
    }

    /**
     * Update the order of the labels inside the database based on the
     * rearrangement that was done inside the adapter.
     */
    override fun onLabelRearranged() {
        for (i in labels.indices) {
            viewModel.editLabelOrder(labels[i].title, i)
        }
    }

    private fun updateRecyclerViewVisibility() {
        if (adapter.itemCount == 0) {
            recyclerView.visibility = View.GONE
            emptyState.visibility = View.VISIBLE
        } else {
            recyclerView.visibility = View.VISIBLE
            emptyState.visibility = View.GONE
        }
    }

    private fun addLabel() {
        labelToAdd =
            Label(addLabelView.text.toString().trim { it <= ' ' }, labelToAdd.colorId)
        if (labelIsGoodToAdd(this, labels, labelToAdd.title, "")) {
            labels.add(labelToAdd)
            adapter.notifyItemInserted(labels.size)
            recyclerView.scrollToPosition(labels.size - 1)
            viewModel.addLabel(labelToAdd)
            labelToAdd = Label(
                "", ThemeHelper.getColor(
                    this,
                    ThemeHelper.COLOR_INDEX_UNLABELED
                )
            )
            addLabelView.setText("")
        }
    }

    companion object {
        private val TAG = AddEditLabelActivity::class.java.simpleName

        /**
         * Used for checking label names before adding or renaming existing ones.
         * Checks for invalid strings like empty ones, spaces or duplicates
         * @param newLabel The desired name for the new label or renamed label
         * @return true if the label name is valid, false otherwise
         */
        @JvmStatic
        fun labelIsGoodToAdd(
            context: Context,
            labels: List<Label>,
            newLabel: String,
            beforeEdit: String
        ): Boolean {
            var result = true
            if (beforeEdit != "" && beforeEdit == newLabel) {
                result = false
            } else if (newLabel.isEmpty()) {
                result = false
            } else {
                var duplicateFound = false
                for (l in labels) {
                    if (newLabel == l.title) {
                        duplicateFound = true
                        break
                    }
                }
                if (duplicateFound) {
                    Toast.makeText(context, R.string.label_already_exists, Toast.LENGTH_SHORT)
                        .show()
                    result = false
                }
            }
            return result
        }
    }
}